lib/rucio/core/replica.py

```diff
@@ -2508,115 +2694,142 @@ def get_and_lock_file_replicas_for_dataset(scope, name, nowait=False, restrict_r
+        )
 
         if restrict_rses is not None:
             if len(restrict_rses) < 10:
                 rse_clause = []
                 for rse_id in restrict_rses:
                     rse_clause.append(models.RSEFileAssociation.rse_id == rse_id)
                 if rse_clause:
-                    query = session.query(models.DataIdentifierAssociation.child_scope,
-                                          models.DataIdentifierAssociation.child_name,
-                                          models.DataIdentifierAssociation.bytes,
-                                          models.DataIdentifierAssociation.md5,
-                                          models.DataIdentifierAssociation.adler32,
-                                          models.RSEFileAssociation)\
-                                   .with_hint(models.DataIdentifierAssociation,
-                                              "INDEX_RS_ASC(CONTENTS CONTENTS_PK) NO_INDEX_FFS(CONTENTS CONTENTS_PK)",
-                                              'oracle')\
-                                   .filter(and_(models.DataIdentifierAssociation.child_scope == models.RSEFileAssociation.scope,
-                                                models.DataIdentifierAssociation.child_name == models.RSEFileAssociation.name,
-                                                models.RSEFileAssociation.state != ReplicaState.BEING_DELETED,
-                                                or_(*rse_clause)))\
-                                   .filter(models.DataIdentifierAssociation.scope == scope,
-                                           models.DataIdentifierAssociation.name == name)
+
+                    stmt = select(
+                        models.DataIdentifierAssociation.child_scope,
+                        models.DataIdentifierAssociation.child_name,
+                        models.DataIdentifierAssociation.bytes,
+                        models.DataIdentifierAssociation.md5,
+                        models.DataIdentifierAssociation.adler32,
+                        models.RSEFileAssociation
+                    ).with_hint(
+                        models.DataIdentifierAssociation,
+                        'INDEX_RS_ASC(CONTENTS CONTENTS_PK) NO_INDEX_FFS(CONTENTS CONTENTS_PK)', 'oracle'
+                    ).with_hint(
+                        models.RSEFileAssociation, 'INDEX(REPLICAS REPLICAS_PK)', 'oracle'
+                    ).where(
+                        and_(models.DataIdentifierAssociation.child_scope == models.RSEFileAssociation.scope,
+                             models.DataIdentifierAssociation.child_name == models.RSEFileAssociation.name,
+                             models.RSEFileAssociation.state != ReplicaState.BEING_DELETED,
+                             models.DataIdentifierAssociation.scope == scope,
+                             models.DataIdentifierAssociation.name == name,
+                             or_(*rse_clause))
+                    )
 
     else:
```

Comment:
Would it make more sense to make a base query and just change the filters with the if/else? None of the base queries the DataIdentifierAssociation Table seem to be that different between any of the parts of the control loop in this function.

Reply:
Good call, refactored the queries in this function (`get_and_lock_file_replicas_for_dataset`) quite a bit. Now using the following two base queries and changing them slightly in the control loop

```
    base_stmt = select(
        models.DataIdentifierAssociation.child_scope,
        models.DataIdentifierAssociation.child_name,
        models.DataIdentifierAssociation.bytes,
        models.DataIdentifierAssociation.md5,
        models.DataIdentifierAssociation.adler32,
    ).where(
        and_(models.DataIdentifierAssociation.scope == scope,
             models.DataIdentifierAssociation.name == name)
    )

    stmt = base_stmt.add_columns(
        models.RSEFileAssociation
    ).where(
        and_(models.DataIdentifierAssociation.child_scope == models.RSEFileAssociation.scope,
             models.DataIdentifierAssociation.child_name == models.RSEFileAssociation.name,
             models.RSEFileAssociation.state != ReplicaState.BEING_DELETED)
    )
```